core/lib/zksync_core/src/sync_layer/gossip/buffered/mod.rs

```diff
@@ -0,0 +1,320 @@
+//! Buffered [`BlockStore`] implementation.
+
+use async_trait::async_trait;
+
+use std::{collections::BTreeMap, ops, time::Instant};
+
+#[cfg(test)]
+use zksync_concurrency::ctx::channel;
+use zksync_concurrency::{
+    ctx, scope,
+    sync::{self, watch, Mutex},
+};
+use zksync_consensus_roles::validator::{BlockNumber, FinalBlock};
+use zksync_consensus_storage::{BlockStore, StorageError, StorageResult, WriteBlockStore};
+
+#[cfg(test)]
+mod tests;
+
+use super::{
+    metrics::{BlockResponseKind, METRICS},
+    utils::MissingBlockNumbers,
+};
+
+/// [`BlockStore`] variation that upholds additional invariants as to how blocks are processed.
+///
+/// The invariants are as follows:
+///
+/// - Stored blocks always have contiguous numbers; there are no gaps.
+/// - Blocks can be scheduled to be added using [`Self::schedule_next_block()`] only. New blocks do not
+///   appear in the store otherwise.
+#[async_trait]
+pub(super) trait ContiguousBlockStore: BlockStore {
+    /// Schedules a block to be added to the store. Unlike [`WriteBlockStore::put_block()`],
+    /// there is no expectation that the block is added to the store *immediately*. It's
+    /// expected that it will be added to the store eventually, which will be signaled via
+    /// a subscriber returned from [`BlockStore::subscribe_to_block_writes()`].
+    ///
+    /// [`Buffered`] guarantees that this method will only ever be called:
+    ///
+    /// - with the next block (i.e., one immediately after [`BlockStore::head_block()`])
+    /// - sequentially (i.e., multiple blocks cannot be scheduled at once)
+    async fn schedule_next_block(&self, ctx: &ctx::Ctx, block: &FinalBlock) -> StorageResult<()>;
+}
+
+#[derive(Debug)]
+struct BlockBuffer {
+    store_block_number: BlockNumber,
+    blocks: BTreeMap<BlockNumber, FinalBlock>,
+}
+
+impl BlockBuffer {
+    fn new(store_block_number: BlockNumber) -> Self {
+        Self {
+            store_block_number,
+            blocks: BTreeMap::new(),
+        }
+    }
+
+    fn head_block(&self) -> Option<FinalBlock> {
+        self.blocks.values().next_back().cloned()
+    }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    fn set_store_block(&mut self, store_block_number: BlockNumber) {
+        assert!(
+            store_block_number > self.store_block_number,
+            "`ContiguousBlockStore` invariant broken: unexpected new head block number"
+        );
```

Comment:
Same qestion here and further down the code: is DoS possible?

Reply:
No, this assertion just encapsulates the `ContiguousBlockStore` invariant (blocks are saved in order of their numbers; this is true for the state keeper now and in the observable future – even if we somehow manage to parallelize transaction processing within a block, we'd still probably execute and persist blocks in order). The external factors like node peers cannot influence this invariant; if it's broken, it means that there's a bug in the node implementation.